.pre-commit-config.yaml

```diff
@@ -3,36 +3,57 @@ default_language_version:
   python: "3.9"
 repos:
   - repo: https://github.com/pre-commit/pre-commit-hooks
-    rev: v2.3.0
+    rev: "v4.1.0"
     hooks:
       - id: check-yaml
+        exclude: .clang-format
       - id: end-of-file-fixer
       - id: trailing-whitespace
+
   - repo: https://github.com/psf/black
-    rev: 20.8b1
+    rev: "22.1.0"
     hooks:
       - id: black
+
   - repo: https://gitlab.com/pycqa/flake8
-    rev: 3.7.8
+    rev: 3.9.2
     hooks:
       - id: flake8
         types: [file, python]
         # only check for unused imports, as the rest is done by black
         args: [--select=F401]
+
   - repo: https://github.com/pre-commit/mirrors-mypy
-    rev: v0.812
+    rev: v0.931
     hooks:
       - id: mypy
         args:
+          - --show-error-codes
           - --ignore-missing-imports
-  - repo: local
+        files: ^(src/|packages/.*/test|conftest.py|docs/)
+        additional_dependencies: &mypy-deps
+          - packaging
+          - types-docutils
+          - types-pyyaml
+          - types-setuptools
+      - id: mypy
+        name: mypy-micropip
+        files: ^packages/micropip/src
+        args: &mypy-args
+          - --show-error-codes
+        additional_dependencies: *mypy-deps
+      - id: mypy
+        name: mypy-pyodide-build
```

Comment:
Do you think it's better to separate these by package, as opposed to running them together?

Reply:
That's what I did. And that's how you solve the dual `setup.py` bug. Since the environments exactly match, pre-commit caches them together.